REGR: concat of unaligned empty DataFrames failing

While trying to get dask's CI passing (https://github.com/dask/dask/pull/6996/), I noticed another error related to concat. Dask concatenates the empty "meta" dataframe to know the shape/dtypes of the resulting dataframe, and something is failing in there now. 

Small reproducer without dask:

```python
>>> df1 = pd.DataFrame({'a': [1, 2, 3], 'b': [4, 5, 6]})
>>> df2 = pd.DataFrame({'b': [1, 2, 3], 'c': [4, 5, 6]})

>>> pd.concat([df1[0:0], df2[0:0], df1[0:0]])
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-151-b1206da866d9> in <module>
----> 1 pd.concat([df1[0:0], df2[0:0], df1[0:0]])

~/scipy/pandas/pandas/core/reshape/concat.py in concat(objs, axis, join, ignore_index, keys, levels, names, verify_integrity, sort, copy)
    297     )
    298 
--> 299     return op.get_result()
    300 
    301 

~/scipy/pandas/pandas/core/reshape/concat.py in get_result(self)
    518 
    519             new_data = concatenate_block_managers(
--> 520                 mgrs_indexers, self.new_axes, concat_axis=self.bm_axis, copy=self.copy
    521             )
    522             if not self.copy:

~/scipy/pandas/pandas/core/internals/concat.py in concatenate_block_managers(mgrs_indexers, axes, concat_axis, copy)
     89         else:
     90             b = make_block(
---> 91                 _concatenate_join_units(join_units, concat_axis, copy=copy),
     92                 placement=placement,
     93                 ndim=len(axes),

~/scipy/pandas/pandas/core/internals/concat.py in _concatenate_join_units(join_units, concat_axis, copy)
    325         join_units = nonempties
    326 
--> 327     empty_dtype, upcasted_na = _get_empty_dtype_and_na(join_units)
    328 
    329     to_concat = [

~/scipy/pandas/pandas/core/internals/concat.py in _get_empty_dtype_and_na(join_units)
    436 
    437     msg = "invalid dtype determination in get_concat_dtype"
--> 438     raise AssertionError(msg)
    439 
    440 

AssertionError: invalid dtype determination in get_concat_dtype
```

It occurs when reindexing happens (not fully aligned dataframes), and apparently at least 3 dataframes are needed to trigger it (the same example with only 2 dataframes passed to concat doesn't fail).

I suppose this might be related to https://github.com/pandas-dev/pandas/pull/38843 (change only on master, so not a target for 1.2.1) cc @jbrockmendel

--- a/pandas/tests/reshape/concat/test_empty.py
+++ b/pandas/tests/reshape/concat/test_empty.py
@@ -263,3 +263,12 @@ class TestEmptyConcat:
         result = concat([df3, df4])
         expected = DataFrame(columns=["a", "b"])
         tm.assert_frame_equal(result, expected)
+
+    def test_concat_empty_dataframe_different_dtypes(self):
+        # 39037
+        df1 = DataFrame({'a': [1, 2, 3], 'b': ['a', 'b', 'c']})
+        df2 = DataFrame({'a': [1, 2, 3]})
+
+        result = concat([df1[:0], df2[:0]])
+        assert result["a"].dtype == np.int64
+        assert result["b"].dtype == np.object_

--- a/pandas/tests/reshape/concat/test_empty.py
+++ b/pandas/tests/reshape/concat/test_empty.py
@@ -249,3 +249,20 @@ class TestEmptyConcat:
         result = concat([df1, df2])
         expected = df1.dtypes
         tm.assert_series_equal(result.dtypes, expected)
+
+    def test_concat_empty_dataframe(self):
+        # 39037
+        df1 = DataFrame({'a': [1, 2, 3], 'b': [4, 5, 6]})
+        df2 = DataFrame({'b': [1, 2, 3], 'c': [4, 5, 6]})
+        result = concat([df1[0:0], df2[0:0], df1[0:0]]).astype('int32')
+        expected = DataFrame(columns=['a', 'b', 'c']).reset_index(drop=True).astype('int32')
+        tm.assert_frame_equal(result, expected)
+    
+    def test_concat_empty_dataframe_dtypes(self):
+        # 39037
+        df1 = DataFrame({'a': [1, 2, 3], 'b': ['a', 'b', 'c']})
+        df2 = DataFrame({'a': [1, 2, 3]})
+
+        result = concat([df1[:0], df2[:0]])
+        assert result["a"].dtype == np.int64
+        assert result["b"].dtype == np.object

--- a/pandas/tests/reshape/concat/test_empty.py
+++ b/pandas/tests/reshape/concat/test_empty.py
@@ -255,12 +255,11 @@ class TestEmptyConcat:
         df1 = DataFrame(columns=["a", "b"])
         df2 = DataFrame(columns=["b", "c"])
         result = concat([df1, df2, df1])
-        expected = DataFrame(columns=['a', 'b', 'c'])
+        expected = DataFrame(columns=["a", "b", "c"])
         tm.assert_frame_equal(result, expected)
 
         df3 = DataFrame(columns=["a", "b"])
         df4 = DataFrame(columns=["b"])
         result = concat([df3, df4])
-        expected = DataFrame(columns=['a', 'b'])
-        tm.assert_frame_equal(result, expected)   
-
+        expected = DataFrame(columns=["a", "b"])
+        tm.assert_frame_equal(result, expected)

--- a/pandas/tests/reshape/concat/test_empty.py
+++ b/pandas/tests/reshape/concat/test_empty.py
@@ -258,7 +258,7 @@ class TestEmptyConcat:
         expected = DataFrame(columns=['a', 'b', 'c']).reset_index(drop=True).astype('int32')
         tm.assert_frame_equal(result, expected)
     
-    def test_concat_empty_dataframe_dtypes(self):
+    def test_concat_empty_dataframe_different_dtypes(self):
         # 39037
         df1 = DataFrame({'a': [1, 2, 3], 'b': ['a', 'b', 'c']})
         df2 = DataFrame({'a': [1, 2, 3]})

--- a/pandas/tests/reshape/concat/test_empty.py
+++ b/pandas/tests/reshape/concat/test_empty.py
@@ -252,17 +252,15 @@ class TestEmptyConcat:
 
     def test_concat_empty_dataframe(self):
         # 39037
-        df1 = DataFrame({'a': [1, 2, 3], 'b': [4, 5, 6]})
-        df2 = DataFrame({'b': [1, 2, 3], 'c': [4, 5, 6]})
-        result = concat([df1[0:0], df2[0:0], df1[0:0]]).astype('int32')
-        expected = DataFrame(columns=['a', 'b', 'c']).reset_index(drop=True).astype('int32')
+        df1 = DataFrame(columns=["a", "b"])
+        df2 = DataFrame(columns=["b", "c"])
+        result = concat([df1, df2, df1])
+        expected = DataFrame(columns=['a', 'b', 'c'])
         tm.assert_frame_equal(result, expected)
-    
-    def test_concat_empty_dataframe_different_dtypes(self):
-        # 39037
-        df1 = DataFrame({'a': [1, 2, 3], 'b': ['a', 'b', 'c']})
-        df2 = DataFrame({'a': [1, 2, 3]})
 
-        result = concat([df1[:0], df2[:0]])
-        assert result["a"].dtype == np.int64
-        assert result["b"].dtype == np.object
+        df3 = DataFrame(columns=["a", "b"])
+        df4 = DataFrame(columns=["b"])
+        result = concat([df3, df4])
+        expected = DataFrame(columns=['a', 'b'])
+        tm.assert_frame_equal(result, expected)   
+

--- a/pandas/tests/reshape/concat/test_empty.py
+++ b/pandas/tests/reshape/concat/test_empty.py
@@ -266,8 +266,8 @@ class TestEmptyConcat:
 
     def test_concat_empty_dataframe_different_dtypes(self):
         # 39037
-        df1 = DataFrame({'a': [1, 2, 3], 'b': ['a', 'b', 'c']})
-        df2 = DataFrame({'a': [1, 2, 3]})
+        df1 = DataFrame({"a": [1, 2, 3], "b": ["a", "b", "c"]})
+        df2 = DataFrame({"a": [1, 2, 3]})
 
         result = concat([df1[:0], df2[:0]])
         assert result["a"].dtype == np.int64

